File: classic/Code/LongestConsecutiveSequence.java
public class Solution {
    public static int longestConsecutive(int[] num) {
        if(num.length==0){
            return 0;
        }
        Set<Integer>set=new HashSet<Integer>();
        for (int i:num){
            set.add(i);
        }
        int m=0;
        for(int e:num){
            int left=e-1;
            int right=e+1;
            int counter=1;
            while(set.contains(left)){
                counter++;
                set.remove(left);
                left--;
            }
            while(set.contains(right)){
                counter++;
                set.remove(right);
                right++;
            }
            m=Math.max(counter,m);
        }
        return m;
    }
}

File: classic/Code/LongestConsecutiveSequence.cpp
class Solution {
public:
    int consecutiveLength(int k,unordered_map<int,int>&hashmap,vector<int>&length){
        if(hashmap.find(k)==hashmap.end()){
            return 0;
        }
        int i=hashmap[k];
        if(length[i]>0){
            return length[i];
        }
        length[i]=consecutiveLength(k-1,hashmap,length)+1;
        return length[i];
    }
    int longestConsecutive(vector<int> &num) {
        unordered_map<int,int>hashmap;
        vector<int>length(num.size(),0);
        for(int i=0;i<num.size();i++){
            hashmap[num[i]]=i;
        }
        for(int i=0;i<num.size();i++){
            if(length[i]>0)continue;
            length[i]=consecutiveLength(num[i],hashmap,length);
        }
        int m=0;
        for(int i=0;i<num.size();i++){
            m=max(m,length[i]);
        }
        return m;
    }
};
